### ISSUE TITLE
Insert Non-StringFields fields in a Form

### ISSUE THREAD

**Issue by leovamos** [volkanceylan/Serenity#4622]
Hi guys,

I have the following problem: my project has some rows which none of the fields are String. And I need to show a field for these rows that are Stream (it is binary in SQL), and byte array (varbinary in SQL) in, for example, a form.

Showing the codes:

`
namespace DashboardSMICv2.Smic.Entities
{
    using Serenity;
    using Serenity.ComponentModel;
    using Serenity.Data;
    using Serenity.Data.Mapping;
    using System;
    using System.ComponentModel;
    using System.IO;
    using System.Text;

    [ConnectionKey("Smic"), Module("Smic"), TableName("[dbo].[AppCertificateHash]")]
    [DisplayName("App Certificate Hash"), InstanceName("App Certificate Hash")]
    [ReadPermission("Administration:General")]
    [ModifyPermission("Administration:General")]
    [LookupScript("Smic.AppCertificateHash")]
    public sealed class AppCertificateHashRow : Row, IIdRow
    {
        [DisplayName("App"), ForeignKey("[dbo].[App]", "AppId"), LeftJoin("jApp"), TextualField("AppUniqueName")]
        public Int32? AppId
        {
            get { return Fields.AppId[this]; }
            set { Fields.AppId[this] = value; }
        }
        [Identity,PrimaryKey]
        public Int32? HashId
        {
            get { return Fields.HashId[this]; }
            set { Fields.HashId[this] = value; }
        }

        [DisplayName("Certificate Hash"), Size(32),LookupInclude]
        public Stream CertificateHash
        {
            get { return Fields.CertificateHash[this]; }
            set { Fields.CertificateHash[this] = value; }
        }
`
The field of interest is CertificateHash

`
namespace DashboardSMICv2.Smic.Entities
{
    using Serenity;
    using Serenity.ComponentModel;
    using Serenity.Data;
    using Serenity.Data.Mapping;
    using System;
    using System.ComponentModel;
    using System.IO;

    [ConnectionKey("Smic"), Module("Smic"), TableName("[dbo].[AppExtra]")]
    [DisplayName("App Extra"), InstanceName("App Extra")]
    [ReadPermission("Administration:General")]
    [ModifyPermission("Administration:General")]
    [LookupScript("Smic.AppExtra")]
    public sealed class AppExtraRow : Row, IIdRow
    {
        [DisplayName("App"), ForeignKey("[dbo].[App]", "AppId"), LeftJoin("jApp"), TextualField("AppUniqueName")]
        public Int32? AppId
        {
            get { return Fields.AppId[this]; }
            set { Fields.AppId[this] = value; }
        }

        [Identity,PrimaryKey]
        public Int32? ExtraId
        {
            get { return Fields.ExtraId[this]; }
            set { Fields.ExtraId[this] = value; }
        }

        [DisplayName("Key"), Size(16), NotNull, LookupInclude]
        public byte[] Key
        {
            get { return Fields.Key[this]; }
            set { Fields.Key[this] = value; }
        }

        [DisplayName("Value"), Size(255), NotNull, LookupInclude]
        public byte[] Value
        {
            get { return Fields.Value[this]; }
            set { Fields.Value[this] = value; }
        }

`

The fields of interest are Key and Value.

In short, what I want to know is how to display these non-StringField fields in a form, because currently I have been able to display them on the grid through dialog.ts of rows where needed, but displaying in a form is being problematic, because they are even listed in the dropdown, but they appear blank and I can't select them. Some screenshots of the scene:
![smic_help_form](https://user-images.githubusercontent.com/38430921/65540479-e2777c00-dee1-11e9-9425-b348130d3ac1.png)

![smic_help_form2](https://user-images.githubusercontent.com/38430921/65540487-e73c3000-dee1-11e9-9931-f121ee7851c2.png)

I know NameField solves this, but since these fields are not String, the INameRow interface is not even implemented in these Rows, so I wanted to know a way around this.

**Comment by JohnRanger** [comment]
Hi @leovamos ,

I gave you already some hints like doing the conversion directly in the database on the fly (with calculated fields). Have you already tried this?

Which Database system (MS SQL, MySQL, etc) are you using?

John

**Comment by leovamos** [comment]
Hi @JohnRanger ,

Apparently the people I work with say that I can't have a calculated field in the database, although I think it's a good idea too. And seeing some methods in Serenity, there is one that converts byte array to string (Convert.ToBase64String), but it converts to type "string" from the System namespace, and the Serenity NameField must be "StringField".
I'm close to the solution for that, but there is still a problem...

**Comment by JohnRanger** [comment]
Hi @leovamos ,

are they allowing you to have an additional view?

Then this could be for you: https://stackoverflow.com/questions/3289988/varbinary-to-string-on-sql-server

And if they don't allow, you probably can do this directly in your serenity code by using an expression on the virtual field (field only defined within your SF project, not in DB) - and the load would still be on the SQL Server :-)

With kind regards,

John

**Comment by leovamos** [comment]
I would like to do this in Serenity itself because I can't change the DB as I want. But I don't know where to insert this code, and even where I start... If you give me a light (again), it would be great.

Thanks in advance

**Comment by JohnRanger** [comment]
Hi @leovamos ,

create an additional field in your AppCertificateHashRow.cs - like this:

```cs
        [DisplayName("CertificateHash Converted"), LookupInclude]
        [Expression("(select CONVERT(varchar(max), T0.CertificateHash, 0))")]
        public String CertificateHashConverted
        {
            get { return Fields.CertificateHashConverted[this]; }
            set { Fields.CertificateHashConverted[this] = value; }
        }
```

also within your AppCertificateHashRow.cs, change your Namefield like this:

```cs
        StringField INameRow.NameField
        {
            get { return Fields.CertificateHashConverted; }
        }
```

Then do a full rebuild and a T4 transform.

Although I couldn't test this here, it should now fill the dropdown visibly with the hashes.

Hope this helps.

With kind regards,

John

**Comment by leovamos** [comment]
Hi @JohnRanger 

Thanks for your help. A colleague helped me with the same solution you proposed. Many thanks, John, you're amazing.

**Comment by JohnRanger** [comment]
HI @leovamos ,
no - I am not :-)  I didn't even know beforehand that this would work. But I have seen in the past people doing some SQL stuff within the Expresison tag within Serenity - and so I just searched for the plain MS SQL solution to convert varbinary to varchar - and then with a bit of trial and error put a working conversion together within xyzRow.cs.

And then the rest is easy - just adjust the namefield to be this new virtual field and by the really -really amazing clever mechanisms of Serenity, this automatically fills the dropdown correctly.

This is a bit like putting together a puzzle :-)

I have learned something by doing this.

Regards,

John
